File: codes/sola.m
function [ysn_sola] = sola(y,frSize,frShift,win,tsmf,max_delay,plot_crosscorr)

% Purpose : To implement SOLA method for time-scale modification fo speech
% signals

% Inputs:
% [1] y - original signal
% [2] frSize - frame size in samples (usually 30ms*sr)
% [3] frShift - frame shift in samples (usually 10ms*sr)
% [4] win - hamming window of length frame size
% [5] tsmf - time-scale modification factor (0.5 to 1.3)
% [6] max_delay - maximum allowable delay for cross-correlation
% [7] plot_crosscorr - Plotting and visualizing the details of SOLA steps 

% Outputs:
% [1] ysn_sola - sola modified signal

% References
% [1] S. Roucos, A. Wilgus, ‘High quality Time-Scale Modification of Speech’, ICASSP-85, pp. 236-239, 1985.


% Author : Sivanand Achanta


% Set analysis and synthesis frame anchor points
len_synsig = ceil(tsmf*length(y));
ati = 0:frShift:length(y)-frSize; % analysis time instants
sti = 0:(tsmf*frShift):(len_synsig)-frSize; % synthesis time instants
v = win.^2;

% Initialization of Sola signal and normalization signal
ys_sola = zeros(len_synsig,1);
normalization_sig = zeros(len_synsig,1) + 1e-6;
delay = 0;

if plot_crosscorr
    h1 = figure(1);
end

for i = 1:length(sti)
    
    sp_nodelay = sti(i) + 1;
    ep_nodelay = sti(i) + frSize;
    
    if (i > 1)
        
        % Take the windowed analysis segment
        x_w = win.*y(ati(i)+1:ati(i)+frSize);
        
        % Do the cross-correlation in both directions (Actually one of them is redundant will remove in the next revision)
        r1 = xcorr(x_w,ys_sola(sp_nodelay:ep_nodelay));
        r2 = xcorr(ys_sola(sp_nodelay:ep_nodelay),x_w);
        
        % Compute the maximum correlation and delay in each of the direction
        [maxval1,d1] = max(r1(frSize:frSize+max_delay));
        d1 = d1 - 1;
        [maxval2,d2] = max(r2(frSize:frSize+max_delay));
        d2 = d2 - 1;
        
        % Take the delay which gives maximum correlation
        if (maxval1 > maxval2)
            delay = -d1; % note the minus sign here ... ***
        else
            delay = d2;
        end
        
        % Plot
        if plot_crosscorr
            nr = 6; nc = 1;
            
            % Plot the signal along with the current analysis window
            k = 1; ax(k) = subplot(nr,nc,k); plot((1:ati(i)+frSize),y(1:ati(i)+frSize)); hold on; plot((ati(i)+1:ati(i)+frSize),win,'r-'); hold off; axis tight;
            title('Signal along with the current analysis window','FontSize',12,'FontWeight','bold');
            
            % Place the analysis segment at the synthesis time instant
            k = k + 1; ax(k) = subplot(nr,nc,k); plot([zeros(sti(i),1);x_w]); axis tight;
            title('Analysis segment placed at the synthesis time instant','FontSize',12,'FontWeight','bold');
            
            % Compensate the delay required for synchronization
            k = k + 1; ax(k) =  subplot(nr,nc,k); plot([zeros(sti(i)+delay,1);x_w]);
            title('Delay compensated synthesis time instant','FontSize',12,'FontWeight','bold');
            
            % Plot the cross correlation fucntion in both directions
            k = k + 1; ax(k) =  subplot(nr,nc,k); plot(r1); axis tight;
            title('Cross-correlation in one direction','FontSize',12,'FontWeight','bold');
            
            k = k + 1; ax(k) =  subplot(nr,nc,k); plot(r2); axis tight;
            title('Cross-correlation in other direction','FontSize',12,'FontWeight','bold');
            
            % Plot the SOLA signal constructed so far
            k = k + 1; ax(k) =  subplot(nr,nc,k); plot(ys_sola(1:ep_nodelay)); axis tight;
            title('SOLA signal constructed so far','FontSize',12,'FontWeight','bold');
            
            linkaxes(ax(1:3),'x');
            xlabel('Samples','FontSize',12,'FontWeight','bold');
            pause
        end
        
    else
        x_w = win.*y(ati(i)+1:ati(i)+frSize);
    end
    
    % Add the delay (here SOLA differs from OLA method)
    sp = sti(i) + 1 + delay ;
    ep = sti(i) + frSize + delay ;
    
    % OLA after synchronization (same as OLA)
    ys_sola(sp:ep) = ys_sola(sp:ep) + win.*x_w;
    normalization_sig(sp:ep) = normalization_sig(sp:ep) + v;
end

ysn_sola = ys_sola./normalization_sig;

File: codes/sola.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 25 11:25:38 2016

@author: Sivanand Achanta
"""

import numpy as np


def ola(y, w, frSize, frShift, tsmf):

    len_synsig = int(np.floor(tsmf*len(y[0])))

    ati = np.floor(np.arange(0, len(y[0]) - frSize, frShift)).astype('int')  # analysis time instants
    sti = np.floor(np.arange(0, len_synsig - frSize, int(tsmf*frShift))).astype('int')  # synthesis time instants

    if len(sti) > len(ati):
        sti = sti[:-1]

    v = w**2

    ys_ola = np.zeros((len_synsig, 1))
    norm_sig = np.zeros((len_synsig, 1)) + 1e-6

    for ti in range(len(sti)):

        sp = int(sti[ti])
        ep = int(sti[ti] + frSize)

        x_w = w*y[0][int(ati[ti]):int(ati[ti] + frSize)]

        # OLA signal
        ys_ola[sp:ep] = ys_ola[sp:ep] + ((w.reshape(1, len(w)))*x_w).T

        # Normalization signal
        norm_sig[sp:ep] = norm_sig[sp:ep] + v.reshape(len(v), 1)

    ysn_ola = ys_ola/norm_sig

    return(ysn_ola)


def sola(y, w, frSize, frShift, tsmf, max_delay):

    len_synsig = int(np.ceil(tsmf*len(y[0])))
    ati = np.floor(np.arange(0, len(y[0]) - frSize, frShift)).astype('int')  # analysis time instants
    sti = np.floor(np.arange(0, len_synsig - frSize, tsmf*frShift)).astype('int')  # synthesis time instants

    if len(sti) > len(ati):
        sti = sti[:-1]

    v = w**2
    delay = 0

    ys_sola = np.zeros((len_synsig, 1))
    norm_sig = np.zeros((len_synsig, 1)) + 1e-6

    for ti in range(len(sti)):
        sp_nodelay = int(sti[ti])
        ep_nodelay = int(sti[ti] + frSize)

        x_w = w*y[0][int(ati[ti]):int(ati[ti] + frSize)]

        if ti > 0:
            r1 = np.correlate(x_w.reshape(frSize,), ys_sola[sp_nodelay:ep_nodelay].reshape(frSize,), 'full')
            r2 = np.correlate(ys_sola[sp_nodelay:ep_nodelay].reshape(frSize,), x_w.reshape(frSize,), 'full')

            r1_max = np.max(r1[frSize:frSize+max_delay])
            r2_max = np.max(r2[frSize:frSize+max_delay])

            d1 = np.argmax(r1[frSize:frSize+max_delay])
            d2 = np.argmax(r2[frSize:frSize+max_delay])

            if r1_max > r2_max:
                delay = -d1
            else:
                delay = d2

        sp = int(sti[ti] + delay)
        ep = int(sti[ti] + frSize + delay)

        # OLA signal
        ys_sola[sp:ep] = ys_sola[sp:ep] + ((w.reshape(1, len(w)))*x_w).T

        # Normalization signal
        norm_sig[sp:ep] = norm_sig[sp:ep] + v.reshape(len(v), 1)

    ysn_sola = ys_sola/norm_sig

    return(ysn_sola)


'''
# Debugging OLA methods
fs = 16000
frSizems = 30
frShiftms = 10
frSize = int((frSizems/1000)*fs)
frShift = int((frShiftms/1000)*fs)
frOvlap = frSize - frShift
nfft = 512
nfftby2 = np.round(nfft/2 + 1)
hfhz = np.linspace(0, fs/2, nfftby2)
w = np.hamming(frSize)
max_delay = 80  # SOLA parmaeter

# Plot flags
pflag = 1
plot_crosscorr = 0

# Matlab plot parameters
font_size = 14

# Time scale modification factor
tsmf = 1

# Generate a synthetic signal (pulse train or a rectangular wave)
t_on = 10
t_off = 160 - t_on

pulse = np.concatenate((np.ones((1, t_on)), np.zeros((1, t_off))), axis=1)
y = np.concatenate((np.zeros((1, 40)), np.tile(pulse, 16)), axis=1)


ysn_ola = ola(y, w, frSize, frShift, tsmf)
ysn_sola = sola(y, w, frSize, frShift, tsmf, max_delay)

if pflag:
    import matplotlib.pyplot as plt

    plt.subplot(311)
    plt.plot(y[0])
    plt.xlim(0, len(y[0]))

    plt.subplot(312)
    plt.plot(ysn_ola, 'r-')
    plt.xlim(0, len(ysn_ola))

    plt.subplot(313)
    plt.plot(ysn_sola, 'm-')
    plt.xlim(0, len(ysn_sola))

'''
